pkg/tfbridge/detailed_diff.go

```diff
@@ -41,6 +41,105 @@ func sortedMergedKeys[K cmp.Ordered, V any, M ~map[K]V](a, b M) []K {
 	return keysSlice
 }
 
+func isTypeShapeMismatched(val resource.PropertyValue, propType shim.ValueType) bool {
+	if !isPresent(val) {
+		return false
+	}
+	switch propType {
+	case shim.TypeList:
+		return !val.IsArray()
+	case shim.TypeSet:
+		return !val.IsArray()
+	case shim.TypeMap:
+		return !val.IsObject()
+	default:
+		return false
+	}
+}
+
+func lookupSchemas(
+	path propertyPath, tfs shim.SchemaMap, ps map[string]*info.Schema,
+) (shim.Schema, *info.Schema, error) {
+	schemaPath := PropertyPathToSchemaPath(resource.PropertyPath(path), tfs, ps)
+	return LookupSchemas(schemaPath, tfs, ps)
+}
+
+// walkPropertyValue walks a property value and calls the visitor function for each path in the property value.
+func walkPropertyValue(
+	val resource.PropertyValue, path propertyPath, visitor func(propertyPath, resource.PropertyValue) bool,
+) bool {
+	if !visitor(path, val) {
+		return false
+	}
+
+	switch {
+	case val.IsArray():
+		for i, elVal := range val.ArrayValue() {
+			if !walkPropertyValue(elVal, path.Index(i), visitor) {
+				return false
+			}
+		}
+
+	case val.IsObject():
+		for k, elVal := range val.ObjectValue() {
+			if !walkPropertyValue(elVal, path.Subpath(string(k)), visitor) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func willTriggerReplacement(
+	path propertyPath, rootTFSchema shim.SchemaMap, rootPulumiSchema map[string]*info.Schema,
+) bool {
+	// A change on a property might trigger a replacement if:
+	// - The property itself is marked as ForceNew
+	// - The direct parent property is a collection (list, set, map) and is marked as ForceNew
+	// See pkg/cross-tests/diff_cross_test.go
+	// TestAttributeCollectionForceNew, TestBlockCollectionForceNew, TestBlockCollectionElementForceNew
+	// for a full case study of replacements in TF
+	tfs, ps, err := lookupSchemas(path, rootTFSchema, rootPulumiSchema)
+	if err != nil {
+		return false
+	}
+	if isForceNew(tfs, ps) {
+		return true
+	}
+
+	if len(path) == 1 {
+		return false
+	}
+
+	parent := path[:len(path)-1]
+	tfs, ps, err = lookupSchemas(parent, rootTFSchema, rootPulumiSchema)
+	if err != nil {
+		return false
+	}
+	// Note this is mimicking the TF behaviour, so the effective type is not considered here.
+	if tfs.Type() != shim.TypeList && tfs.Type() != shim.TypeSet && tfs.Type() != shim.TypeMap {
+		return false
+	}
+	return isForceNew(tfs, ps)
+}
+
+func willTriggerReplacementRecursive(
```

Comment:
The name for this function needs to describe what it does, not how its' implemented.